File: next-macos-native/Sources/MooToolNextCore/ElectronPaneSizeImport.swift
import Foundation

/// Maps Electron normalized pane ratios to native absolute widths (Compose-compatible keys).
public enum ElectronPaneSizeImport {
    private static let referenceWidth = 1320.0

    public static func mergeElectronIntoNative(_ electron: [String: [Double]], current: [String: [Double]]) -> [String: [Double]] {
        guard !electron.isEmpty else { return current }
        var result = current

        if let ratios = electron["json-three-pane"], ratios.count >= 3, let total = positiveSum(ratios) {
            let vault = (ratios[0] / total * referenceWidth).clamped(to: 200...320)
            let inspector = (ratios[2] / total * referenceWidth).clamped(to: 240...340)
            result["json"] = [vault, inspector]
        }
        if let ratios = electron["json-two-pane"], ratios.count >= 2, let total = positiveSum(ratios) {
            let vault = (ratios[0] / total * referenceWidth).clamped(to: 200...320)
            let inspector = existingSlot(result: result, current: current, paneKey: "json", index: 1, fallback: 280)
            result["json"] = [vault, inspector]
        }

        mapTwoColumn(electron["http-workspace"], paneKey: "http", minFirst: 180, maxFirst: 320, defaultSecond: 220, into: &result, current: current)
        mapTwoColumn(electron["host-workspace"], paneKey: "host", minFirst: 170, maxFirst: 360, defaultSecond: 220, into: &result, current: current)
        mapTwoColumn(electron["image-library"], paneKey: "image", minFirst: 180, maxFirst: 360, defaultSecond: 240, into: &result, current: current)
        mapTwoColumn(electron["translation-words"], paneKey: "translation", minFirst: 170, maxFirst: 360, defaultSecond: 220, into: &result, current: current)
        mapTwoColumn(electron["ua-parser"], paneKey: "uaParse", minFirst: 300, maxFirst: 480, defaultSecond: 360, into: &result, current: current)
        mapTwoColumn(electron["text-diff"], paneKey: "textDiff", minFirst: 240, maxFirst: 900, defaultSecond: 240, into: &result, current: current)
        mapTwoColumn(electron["network-workspace"], paneKey: "net", minFirst: 340, maxFirst: 900, defaultSecond: 320, into: &result, current: current)
        mapTwoColumn(electron["regex-test"], paneKey: "regex", minFirst: 320, maxFirst: 900, defaultSecond: 290, into: &result, current: current)
        mapTwoColumn(electron["reformat-file"], paneKey: "reformat", minFirst: 260, maxFirst: 900, defaultSecond: 260, into: &result, current: current)
        mapTwoColumn(electron["calculator-panels"], paneKey: "calculator", minFirst: 360, maxFirst: 900, defaultSecond: 320, into: &result, current: current)
        mapTwoColumn(electron["cron-builder"], paneKey: "cron", minFirst: 460, maxFirst: 900, defaultSecond: 340, into: &result, current: current)
        mapIoThree(electron["encode-panes"], paneKey: "encode", minLeft: 240, maxLeft: 700, minMiddle: 120, maxMiddle: 280, into: &result)
        mapIoThree(electron["config-convert"], paneKey: "ymlProperties", minLeft: 240, maxLeft: 700, minMiddle: 110, maxMiddle: 280, into: &result)
        mapIoThree(electron["config-validate"], paneKey: "ymlProperties", minLeft: 240, maxLeft: 700, minMiddle: 110, maxMiddle: 280, into: &result)
        mapTwoColumn(electron["color-board"], paneKey: "colorBoard", minFirst: 240, maxFirst: 700, defaultSecond: 420, into: &result, current: current)
        mapIoThree(electron["protobuf-wire"], paneKey: "protobuf", minLeft: 240, maxLeft: 700, minMiddle: 110, maxMiddle: 280, into: &result)
        mapIoThree(electron["protobuf-convert"], paneKey: "protobuf", minLeft: 240, maxLeft: 700, minMiddle: 110, maxMiddle: 280, into: &result)
        mapIoThree(electron["crypto-symmetric"], paneKey: "crypto", minLeft: 240, maxLeft: 700, minMiddle: 110, maxMiddle: 280, into: &result)
        mapIoThree(electron["crypto-base"], paneKey: "crypto", minLeft: 240, maxLeft: 700, minMiddle: 110, maxMiddle: 280, into: &result)
        mapTwoColumn(electron["crypto-key-pair"], paneKey: "crypto", minFirst: 260, maxFirst: 900, defaultSecond: 260, into: &result, current: current)
        mapTwoColumn(electron["runtime-editor-output"], paneKey: "java", minFirst: 300, maxFirst: 900, defaultSecond: 300, into: &result, current: current)
        mapTwoColumn(electron["quick-note-editor-preview"], paneKey: "quick-note-editor-preview", minFirst: 260, maxFirst: 900, defaultSecond: 260, into: &result, current: current)
        mapTwoColumn(electron["qrcode-generate"], paneKey: "qrCode", minFirst: 300, maxFirst: 900, defaultSecond: 280, into: &result, current: current)
        mapTwoColumn(electron["qrcode-recognize"], paneKey: "qrCode", minFirst: 300, maxFirst: 900, defaultSecond: 280, into: &result, current: current)

        if let ratios = electron["quick-note-tree-replace"], ratios.count >= 3, let total = positiveSum(ratios) {
            let vault = (ratios[0] / total * referenceWidth).clamped(to: 200...320)
            let replace = (ratios[2] / total * referenceWidth).clamped(to: 200...340)
            result["quick-note-tree-replace"] = [vault, replace]
        }
        if let ratios = electron["quick-note-tree-no-replace"], ratios.count >= 2, let total = positiveSum(ratios) {
            let vault = (ratios[0] / total * referenceWidth).clamped(to: 200...320)
            let replace = existingSlot(result: result, current: current, paneKey: "quick-note-tree-no-replace", index: 1, fallback: 240)
            result["quick-note-tree-no-replace"] = [vault, replace]
        }
        if let ratios = electron["quick-note-no-tree-replace"], ratios.count >= 2, let total = positiveSum(ratios) {
            let editor = (ratios[0] / total * referenceWidth).clamped(to: 320...900)
            let replace = (ratios[1] / total * referenceWidth).clamped(to: 200...340)
            result["quick-note-no-tree-replace"] = [editor, replace]
        }

        return result
    }

    private static func positiveSum(_ ratios: [Double]) -> Double? {
        let total = ratios.reduce(0, +)
        return total > 0 ? total : nil
    }

    private static func mapIoThree(_ ratios: [Double]?, paneKey: String, minLeft: Double, maxLeft: Double, minMiddle: Double, maxMiddle: Double, into result: inout [String: [Double]]) {
        guard let ratios, ratios.count >= 3, let total = positiveSum(ratios) else { return }
        let left = (ratios[0] / total * referenceWidth).clamped(to: minLeft...maxLeft)
        let middle = (ratios[1] / total * referenceWidth).clamped(to: minMiddle...maxMiddle)
        result[paneKey] = [left, middle]
    }

    private static func mapTwoColumn(_ ratios: [Double]?, paneKey: String, minFirst: Double, maxFirst: Double, defaultSecond: Double, into result: inout [String: [Double]], current: [String: [Double]]) {
        guard let ratios, ratios.count >= 2, let total = positiveSum(ratios) else { return }
        let first = (ratios[0] / total * referenceWidth).clamped(to: minFirst...maxFirst)
        let second = existingSlot(result: result, current: current, paneKey: paneKey, index: 1, fallback: defaultSecond)
        result[paneKey] = [first, second]
    }

    private static func existingSlot(result: [String: [Double]], current: [String: [Double]], paneKey: String, index: Int, fallback: Double) -> Double {
        if let list = result[paneKey], index >= 0, index < list.count { return list[index] }
        if let list = current[paneKey], index >= 0, index < list.count { return list[index] }
        return fallback
    }
}

private extension Double {
    func clamped(to range: ClosedRange<Double>) -> Double { min(max(self, range.lowerBound), range.upperBound) }
}

File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/domain/ElectronPaneSizeImport.kt
package com.rememberber.mootool.next.compose.domain

import com.rememberber.mootool.next.compose.model.ToolId

/**
 * Electron stores resizable column sizes as normalized ratios under storage keys.
 * Compose stores absolute dp widths per [ToolId] and pane index.
 */
object ElectronPaneSizeImport {
    private const val REFERENCE_CONTENT_WIDTH_DP = 1320f
    private const val REFERENCE_CONTENT_HEIGHT_DP = 900f

    fun mergeElectronIntoCompose(
        electron: Map<String, List<Float>>,
        current: Map<String, List<Float>>
    ): Map<String, List<Float>> {
        if (electron.isEmpty()) return current
        val result = current.toMutableMap()

        electron["json-three-pane"]?.let { ratios ->
            if (ratios.size >= 3) {
                val total = ratios.sum().takeIf { it > 0f } ?: return@let
                val vault = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(200f, 320f)
                val inspector = (ratios[2] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(240f, 340f)
                result[ToolId.Json.id] = listOf(vault, inspector)
            }
        }
        electron["json-two-pane"]?.let { ratios ->
            if (ratios.size >= 2) {
                val total = ratios.sum().takeIf { it > 0f } ?: return@let
                val vault = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(200f, 320f)
                val inspector = existingSlot(result, current, ToolId.Json.id, 1, 280f)
                result[ToolId.Json.id] = listOf(vault, inspector)
            }
        }

        mapTwoColumnWorkspace(electron["http-workspace"], ToolId.Http.id, 180f, 320f, 220f, result, current)
        mapTwoColumnWorkspace(electron["host-workspace"], ToolId.Host.id, 170f, 360f, 220f, result, current)
        mapTwoColumnWorkspace(electron["image-library"], ToolId.Image.id, 180f, 360f, 240f, result, current)
        mapTwoColumnWorkspace(electron["translation-words"], ToolId.Translation.id, 170f, 360f, 220f, result, current)
        mapTwoColumnWorkspace(electron["ua-parser"], ToolId.UaParse.id, 300f, 480f, 360f, result, current)
        mapTwoColumnWorkspace(electron["text-diff"], ToolId.TextDiff.id, 240f, 900f, 240f, result, current)
        mapTwoColumnWorkspace(electron["network-workspace"], ToolId.Net.id, 340f, 900f, 320f, result, current)
        mapTwoColumnWorkspace(electron["regex-test"], ToolId.Regex.id, 320f, 900f, 290f, result, current)
        mapTwoColumnWorkspace(electron["reformat-file"], ToolId.Reformat.id, 260f, 900f, 260f, result, current)
        mapTwoColumnWorkspace(electron["calculator-panels"], ToolId.Calculator.id, 360f, 900f, 320f, result, current)
        mapTwoColumnWorkspace(electron["cron-builder"], ToolId.Cron.id, 460f, 900f, 340f, result, current)
        mapIoThreePane(electron["encode-panes"], ToolId.Encode.id, 240f, 700f, 120f, 280f, result)
        mapIoThreePane(electron["config-convert"], "config-convert", 240f, 700f, 110f, 280f, result)
        mapIoThreePane(electron["config-validate"], "config-validate", 240f, 700f, 110f, 280f, result)
        mapTwoColumnWorkspace(electron["settings-page"], "settings-page", 180f, 360f, 420f, result, current)
        mapTwoColumnWorkspace(electron["color-board"], ToolId.ColorBoard.id, 240f, 700f, 420f, result, current)
        mapTwoColumnWorkspace(electron["translation-editor"], "translation-editor", 280f, 900f, 280f, result, current)
        mapTwoColumnWorkspace(electron["protobuf-json"], "protobuf-json", 240f, 700f, 420f, result, current)
        mapIoThreePane(electron["protobuf-wire"], "protobuf-wire", 240f, 700f, 110f, 280f, result)
        mapIoThreePane(electron["protobuf-convert"], "protobuf-convert", 240f, 700f, 110f, 280f, result)
        mapIoThreePane(electron["crypto-symmetric"], "crypto-symmetric", 240f, 700f, 110f, 280f, result)
        mapIoThreePane(electron["crypto-base"], "crypto-base", 240f, 700f, 110f, 280f, result)
        mapTwoColumnWorkspace(electron["crypto-key-pair"], "crypto-key-pair", 260f, 900f, 260f, result, current)

        mapTwoColumnWorkspace(electron["runtime-editor-output"], "runtime-editor-output", 300f, 900f, 300f, result, current)

        electron["quick-note-tree-replace"]?.let { ratios ->
            if (ratios.size >= 3) {
                val total = ratios.sum().takeIf { it > 0f } ?: return@let
                val vault = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(200f, 320f)
                val replace = (ratios[2] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(200f, 340f)
                result["quick-note-tree-replace"] = listOf(vault, replace)
            }
        }
        electron["quick-note-tree-no-replace"]?.let { ratios ->
            if (ratios.size >= 2) {
                val total = ratios.sum().takeIf { it > 0f } ?: return@let
                val vault = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(200f, 320f)
                val replace = existingSlot(result, current, "quick-note-tree-no-replace", 1, 240f)
                result["quick-note-tree-no-replace"] = listOf(vault, replace)
            }
        }
        electron["quick-note-no-tree-replace"]?.let { ratios ->
            if (ratios.size >= 2) {
                val total = ratios.sum().takeIf { it > 0f } ?: return@let
                val editor = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(320f, 900f)
                val replace = (ratios[1] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(200f, 340f)
                result["quick-note-no-tree-replace"] = listOf(editor, replace)
            }
        }
        mapTwoColumnWorkspace(electron["quick-note-editor-preview"], "quick-note-editor-preview", 260f, 900f, 260f, result, current)
        mapTwoColumnWorkspace(electron["qrcode-generate"], "qrcode-generate", 300f, 900f, 280f, result, current)
        mapTwoColumnWorkspace(electron["qrcode-recognize"], "qrcode-recognize", 300f, 900f, 280f, result, current)

        return result
    }

    private fun mapIoThreePane(
        ratios: List<Float>?,
        paneKey: String,
        minLeft: Float,
        maxLeft: Float,
        minMiddle: Float,
        maxMiddle: Float,
        result: MutableMap<String, List<Float>>
    ) {
        ratios ?: return
        if (ratios.size < 3) return
        val total = ratios.sum().takeIf { it > 0f } ?: return
        val left = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(minLeft, maxLeft)
        val middle = (ratios[1] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(minMiddle, maxMiddle)
        result[paneKey] = listOf(left, middle)
    }

    private fun mapTwoColumnWorkspace(
        ratios: List<Float>?,
        paneKey: String,
        minFirst: Float,
        maxFirst: Float,
        defaultSecond: Float,
        result: MutableMap<String, List<Float>>,
        current: Map<String, List<Float>>
    ) {
        ratios ?: return
        if (ratios.size < 2) return
        val total = ratios.sum().takeIf { it > 0f } ?: return
        val first = (ratios[0] / total * REFERENCE_CONTENT_WIDTH_DP).coerceIn(minFirst, maxFirst)
        val second = existingSlot(result, current, paneKey, 1, defaultSecond)
        result[paneKey] = listOf(first, second)
    }

    private fun existingSlot(
        result: Map<String, List<Float>>,
        current: Map<String, List<Float>>,
        paneKey: String,
        index: Int,
        fallback: Float
    ): Float = result[paneKey]?.getOrNull(index)
        ?: current[paneKey]?.getOrNull(index)
        ?: fallback
}
